evm-contracts/src/v0.6/dev/CheckedMath.sol

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: MIT
+// Adapted from https://github.com/OpenZeppelin/openzeppelin-contracts/blob/97894a140d2a698e5a0f913648a8f56d62277a70/contracts/math/SignedSafeMath.sol
+
+pragma solidity ^0.6.0;
+
+library CheckedMath {
+
+  int256 constant internal INT256_MIN = -2**255;
+
+  /**
+   * @dev Subtracts two signed integers, returns false 2nd param on overflow.
+   */
+  function add(
+    int256 a,
+    int256 b
+  )
+    internal
+    pure
+    returns (int256 result, bool ok)
+  {
+    int256 c = a + b;
+    if ((b >= 0 && c < a) || (b < 0 && c >= a)) return (0, false);
```

Comment:
Hmmm, this expression also evaluates to false:
```
((b >= 0 && c < a) || (b < 0 && c >= a)) =
((b >= 0 && a + b < a) || (b < 0 && a + b >= a)) =
((b >= 0 && b < 0) || (b < 0 && b >= 0)) =
FALSE || FALSE =
FALSE
```

Reply:
only if there is now overflow, which is exactly what we're guarding against here

Reply:
Oh I see! a+b might be larger than 2**256. Now I know what I was missing! Thanks for the explanation!

Reply:
Had a think about this over the weekend. This is a cool technique to detect overflow. You can tell I haven't done much low-level programming. 

It's also scary that you have to do this. I would expect the runtime to trigger an exception for something like this. Say you're using a third-party library, how do you make sure it accounts for all potential overflows?!

Reply:
Agreed that it's scary! 

Some smart contract languages like vyper indeed solve this at the runtime level: https://vyper.readthedocs.io/en/stable/

Third-party libraries in smart contracts are generally tricky. What if there is ANY kind of bug in it?